pkg/auth/auth.go

```diff
@@ -130,6 +137,53 @@ func GetServicePrincipalToken(tenantID, clientID, secret, resource string) (*ada
 	return &token, nil
 }
 
+// GetServicePrincipalTokenWithCertificate return the token for the assigned user with certificate
+func GetServicePrincipalTokenWithCertificate(adEndpointFromSpec, tenantID, clientID string, certificate []byte, password, resource string) (*adal.Token, error) {
+	begin := time.Now()
+	var err error
+
+	defer func() {
+		if err != nil {
+			err = reporter.ReportIMDSOperationError(metrics.AdalTokenOperationName)
+			if err != nil {
+				klog.Warningf("Metrics reporter error: %+v", err)
+			}
+
+			return
+		}
+		err = reporter.ReportIMDSOperationDuration(metrics.AdalTokenOperationName, time.Since(begin))
+		if err != nil {
+			klog.Warningf("Metrics reporter error: %+v", err)
+		}
+	}()
+
+	activeDirectoryEndpoint := defaultActiveDirectoryEndpoint
+	if adEndpointFromSpec != "" {
+		activeDirectoryEndpoint = adEndpointFromSpec
+	}
+	oauthConfig, err := adal.NewOAuthConfig(activeDirectoryEndpoint, tenantID)
+	if err != nil {
+		return nil, fmt.Errorf("creating the OAuth config: %v", err)
```

Comment:
Let's change this error message to `fmt.Errorf("failed to create oauth config, err: %v", err)`